### ISSUE TITLE
timer and monitor wrongly set in MG Configuration

### ISSUE THREAD

**Issue by jkotan** [sardana-org/sardana#1516]
Our users reported a problem with setting the master timer and monitor, i.e. if you sent to MeasurementGroup configuration with timer and monitor which is not on the first place e.g.
```python
mg.Configuration = \
{'controllers': {
  'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp': {'channels': {
    'tango://haso228jk:10000/expchan/sis3820_exp/1': {
     '_controller_name': 'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
     'index': 0,
     'instrument': None,
     'label': 'exp_c01',
     'name': 'exp_c01',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/sis3820_exp/1/Value'},
    'tango://haso228jk:10000/expchan/sis3820_exp/2': {
     '_controller_name': 'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/sis3820_exp/2',
     'index': 1,
     'instrument': None,
     'label': 'exp_c02',
     'name': 'exp_c02',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/sis3820_exp/2/Value'}},
   'synchronization': 0,
   'monitor': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
   'timer': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
   'synchronizer': 'software'},
  'tango://haso228jk:10000/controller/dgg2ctrl/dgg2_exp_02': {'channels': {
    'tango://haso228jk:10000/expchan/dgg2_exp_02/1': {
     '_controller_name': 'tango://haso228jk:10000/controller/dgg2ctrl/dgg2_exp_02',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
     'index': 2,
     'instrument': None,
     'label': 'exp_t02',
     'name': 'exp_t02',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1/Value'}},
   'synchronization': 0,
   'monitor': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
   'timer': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
   'synchronizer': 'software'}},
 'description': 'Measurement Group',
 'label': 'mg_test01',
 'monitor': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
 'timer': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1'}
```
with timer and monitor specified by `{"monitor": "tango://haso228jk:10000/expchan/dgg2_exp_02/1", "timer": "tango://haso228jk:10000/expchan/dgg2_exp_02/1"}`
and then you read back the configuration  you get
```python
ms.Configuration
Out[19]: \
{'controllers': {
  'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp': {
   'synchronizer': 'software',
   'synchronization': 0,
   'channels': {
    'tango://haso228jk:10000/expchan/sis3820_exp/1': {
     '_controller_name': 'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
     'index': 0,
     'instrument': None,
     'label': 'exp_c01',
     'name': 'exp_c01',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/sis3820_exp/1/Value'},
    'tango://haso228jk:10000/expchan/sis3820_exp/2': {
     '_controller_name': 'tango://haso228jk:10000/controller/sis3820ctrl/sis3820_exp',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/sis3820_exp/2',
     'index': 1,
     'instrument': None,
     'label': 'exp_c02',
     'name': 'exp_c02',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/sis3820_exp/2/Value'}},
   'timer': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
   'monitor': 'tango://haso228jk:10000/expchan/sis3820_exp/1'},
  'tango://haso228jk:10000/controller/dgg2ctrl/dgg2_exp_02': {
   'synchronizer': 'software',
   'synchronization': 0,
   'channels': {
    'tango://haso228jk:10000/expchan/dgg2_exp_02/1': {
     '_controller_name': 'tango://haso228jk:10000/controller/dgg2ctrl/dgg2_exp_02',
     '_unit_id': '0',
     'conditioning': '',
     'data_type': 'float64',
     'data_units': '',
     'enabled': True,
     'full_name': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
     'index': 2,
     'instrument': None,
     'label': 'exp_t02',
     'name': 'exp_t02',
     'ndim': 0,
     'nexus_path': '',
     'normalization': 0,
     'output': True,
     'shape': [],
     'plot_axes': ['<mov>'],
     'plot_type': 1,
     'source': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1/Value'}},
   'timer': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1',
   'monitor': 'tango://haso228jk:10000/expchan/dgg2_exp_02/1'}},
 'label': 'mg_test01',
 'description': 'Measurement Group',
 'timer': 'tango://haso228jk:10000/expchan/sis3820_exp/1',
 'monitor': 'tango://haso228jk:10000/expchan/sis3820_exp/1'}
```

so  with `{"timer":"tango://haso228jk:10000/expchan/sis3820_exp/1","monitor":"tango://haso228jk:10000/expchan/sis3820_exp/1" }`

So it is a bug .i.e. the timer and monitor are not properly set 
or inconsistent behavior. i.e. when you set timer/monitor from first item why you need to provide or you reseave back 'timer'  and `monitor`?

We use sardana 3.0.4

**Comment by jkotan** [comment]
Hi,

I've just looked at the current code of `sardana/pool/poolmeasurementgroup.py`  and it seems that 
 if there is any timerable controller in MG it takes the one with the lowest index. 

At DESY we have counters (timerable) by they cannot be treated as timers.
So I think this approach is not so good. The proper code should take the user master timer/monitor and check it exists in timerable. If it exists it should set it as a master timer/monito if not it can take the first timerable.   In the current approach it is also hard to set master timer different then master monitor. 

What do you think?

**Comment by reszelaz** [comment]
Hi @jkotan,
I'm just starting with this issue. Before I formatted the configuration dictionaries (using `json.loads()`) so they fit on one page. Hope you don't mind.

**Comment by reszelaz** [comment]
> I've just looked at the current code of sardana/pool/poolmeasurementgroup.py and it seems that
if there is any timerable controller in MG it takes the one with the lowest index.

Since we are enhancing Sardana with different synchronization modes, we found the concept of master timer/monitor (the one for the whole measurement group) limiting us. For that reason we introduced the master timer/monitor for _software synchronization_ and another one for _software start synchronization_. When refactoring the measurement group configuration while working on SEP18 we decided to keep the _old_ master timer/monitor for backwards compatibility. 

Yes, your intepretation of the code is correct, the selection of master timer/monitor for _software synchronization_ and for _software start synchronization_ is based on the index i.e. order of channels in expconf. Ideally we would like to have more explicit selection in the expconf.

> At DESY we have counters (timerable) by they cannot be treated as timers.
So I think this approach is not so good. The proper code should take the user master timer/monitor and check it exists in timerable. If it exists it should set it as a master timer/monito if not it can take the first timerable.

I agree, to fully maintain backwars compatibiliy we should respect the timer/monitor when setting the configuration. I have just implemented it in #1521. But if I understands well, what is currently in that PR (now it simply returns in the configuration what was passed when settin it) won't be enough. What you actually would like is that the channels set as master timer/monitor are started as the last ones? In this case we would also need to set them as master timer/monitor for the _software synchronization_. If you confirm I will make an additional commit.

> In the current approach it is also hard to set master timer different then master monitor.

Yes, I agree. But this is related to the idea of having the possibility to explicitly select it in the expconf.

**Comment by jkotan** [comment]
Thanks for the answer . Unfortunatelty, with the #1521 I get the following error
```
MainThread     DEBUG    2021-03-13 20:11:57,875 mntgrp/pool_haso228jk/mg_test01: -> init_device
MainThread     DEBUG    2021-03-13 20:11:57,877 mntgrp/pool_haso228jk/mg_test01: <- init_device
MainThread     DEBUG    2021-03-13 20:11:57,878 mntgrp/pool_haso228jk/mg_test02: -> init_device
MainThread     DEBUG    2021-03-13 20:11:57,880 mntgrp/pool_haso228jk/mg_test02: <- init_device
Tango NamedDevFailedList exception
   Exception for object Configuration
   Index of object in call (starting at 0) = 1
       Severity = ERROR 
       Error reason = PyDs_PythonError
       Desc : ValueError: timer tango://haso228jk:10000/expchan/dgg2_exp_02/1 is not present/enabled

       Origin : Traceback (most recent call last):
  File "/home/jkotan/testinst3/lib/python3.7/site-packages/sardana-3.0.4a0-py3.7.egg/sardana/tango/pool/MeasurementGroup.py", line 245, in write_Configuration
    self.measurement_group.set_configuration_from_user(cfg)
  File "/home/jkotan/testinst3/lib/python3.7/site-packages/sardana-3.0.4a0-py3.7.egg/sardana/pool/poolmeasurementgroup.py", line 1113, in set_configuration_from_user
    self._config.set_configuration_from_user(cfg)
  File "/home/jkotan/testinst3/lib/python3.7/site-packages/sardana-3.0.4a0-py3.7.egg/sardana/pool/poolmeasurementgroup.py", line 878, in set_configuration_from_user
    'timer {} is not present/enabled'.format(mnt_grp_timer))
ValueError: timer tango://haso228jk:10000/expchan/dgg2_exp_02/1 is not present/enabled

   Summary exception
       Severity = ERROR 
       Error reason = API_AttributeFailed
       Desc : Failed to execute DeviceClass::set_memorized_values() on device mntgrp/pool_haso228jk/mg_test01, object(s) Configuration
       Origin : DeviceClass::set_memorized_values()

MainThread     INFO     2021-03-13 20:11:57,884 TaurusRootLogger: Ready to accept request in 0:00:01.221449
```
which is not the case for the develop branch.
It looks like the PR does not find all timerable controllers, only the first one of the particular type. Therefore, If find only    expchan/dgg2_exp_01/1 while in the MG we have  expchan/dgg2_exp_02/1

**Comment by jkotan** [comment]
Hi Zibi, 
I think in #1521 the error is caused  by problems in algorithm logic, i.e.
`get_timerable_channels()` is trying to find  timer `tango://haso228jk:10000/expchan/dgg2_exp_02/1`
in its list
```python
        if mnt_grp_timer:
            timerable_channels = self.get_timerable_channels(enabled=True)
            if mnt_grp_timer in [ch.full_name for ch in timerable_channels]:
```
before the timer is added at
```python
        # Update internals values                                                                                                                                       
        self._label = label
        self._description = description
        self._timerable_ctrls = timerable_ctrls
```

**Comment by reszelaz** [comment]
Fixed in #1521
